Grouped write

CourrierGui opened the pull request:
Implementation of a grouped write I/O scheduler to address the need to group requests per tag. This current version does not run requests on different tags in parallel. This will come later.

Commit:
gitignore: add phobos_tlc and doc/man/build

Change-Id: I160547e14ebea306edb877843bc5432df62e41fc

Commit:
io_sched: add skeleton of grouped_write I/O scheduler

Change-Id: I9dd95ef029bd3b79ce23c49bae12a5b7daf9b28f

Commit:
io_sched: add skeleton of grouped_write I/O scheduler

Change-Id: I9dd95ef029bd3b79ce23c49bae12a5b7daf9b28f

Commit:
grouped_write: add request data structures

Write requests will be stored in queues in a hash table.
The queues will be hashed based on the tags and layout
parameters of the write requests. These queues will be
divided into one subqueue per grouping. Groupings will
be written one after the other to increase the data
locality on the tape.

Change-Id: I0db73965f106e751d5e33fda70bd6e3ba6ca463e

Commit:
grouped_write: add request data structures

Write requests will be stored in queues in a hash table.
The queues will be hashed based on the tags and layout
parameters of the write requests. These queues will be
divided into one subqueue per grouping. Groupings will
be written one after the other to increase the data
locality on the tape.

Change-Id: I0db73965f106e751d5e33fda70bd6e3ba6ca463e

Commit:
grouped_write: implement .push_request method

If a queue has already been allocated for this kind of request,
we push it on the queue. Otherwise, we create a new queue.

When pushing to a queue, we try to find the corresponding grouping.
If not found, we create a new one. There is a special case for if
there is no grouping that needs to be managed since the grouping's
name is the key used in the queue's grouping_index hashtable.

Change-Id: Ibd12c52000a19c93c9470089a122158aba0a8b5c

Commit:
grouped_write: implement .push_request method

If a queue has already been allocated for this kind of request,
we push it on the queue. Otherwise, we create a new queue.

When pushing to a queue, we try to find the corresponding grouping.
If not found, we create a new one. There is a special case for if
there is no grouping that needs to be managed since the grouping's
name is the key used in the queue's grouping_index hashtable.

Change-Id: Ibd12c52000a19c93c9470089a122158aba0a8b5c

Commit:
grouped_write: implement .remove_request method

On remove we simply remove the request from its queue.
If this was the last request, we must remove the queue.
When removing the queue, we have to deallocate the devices.

Change-Id: Ie8fa7d3436029a90b9027d92cda3b9aaf40dcb07

Commit:
grouped_write: implement .remove_request method

On remove we simply remove the request from its queue.
If this was the last request, we must remove the queue.
When removing the queue, we have to deallocate the devices.

Change-Id: Ie8fa7d3436029a90b9027d92cda3b9aaf40dcb07

Commit:
grouped_write: implement .{add,remove,get}_device methods

These methods are similar to the fifo implementation.
We simply have to check whether a device was allocated to
a queue before removing it.

Change-Id: Ic83830c590e4cf6f2eb53b5287eef341de5ef374

Commit:
grouped_write: implement .{add,remove,get}_device methods

These methods are similar to the fifo implementation.
We simply have to check whether a device was allocated to
a queue before removing it.

Change-Id: Ic83830c590e4cf6f2eb53b5287eef341de5ef374

Commit:
grouped_write: implementation of .peek_request

This function will simply try to find from which queue
we should pick a request and return that request.

If some devices are free, try to allocate the next queue by moving it to
the allocated list. Then find the first queue whose devices are free to
do an I/O and return the first element of this queue. At this point, the
queue may not have devices associated to it yet.

Change-Id: I2feb13eebd28031fb52e6d1bf517a7b18330acaf

Commit:
grouped_write: implementation of .peek_request

This function will simply try to find from which queue
we should pick a request and return that request.

If some devices are free, try to allocate the next queue by moving it to
the allocated list. Then find the first queue whose devices are free to
do an I/O and return the first element of this queue. At this point, the
queue may not have devices associated to it yet.

Change-Id: I2feb13eebd28031fb52e6d1bf517a7b18330acaf

Commit:
grouped_write: implement .retry and .get_device_medium_pair

Change-Id: Ie476170f796f426030949ba2d61f036bf8e4b009

Commit:
grouped_write: implement .retry and .get_device_medium_pair

Change-Id: Ie476170f796f426030949ba2d61f036bf8e4b009

Commit:
grouped_write: implement .claim_device method

The implementation is very similar to the grouped_read
algorithm.

On exchange, we give back the device if it is not busy.
On take, we give back any device we have available.

Change-Id: I2c99800f25aefd15717ac7e3c53a1b9457b682dd

Commit:
grouped_write: implement .claim_device method

The implementation is very similar to the grouped_read
algorithm.

On exchange, we give back the device if it is not busy.
On take, we give back any device we have available.

Change-Id: I2c99800f25aefd15717ac7e3c53a1b9457b682dd

Commit:
grouped_write: handle .get_device_medium_pair retries

get_device_medium_pair can be called multiple times with the
same index as long as the upper layer is not satisfied with
the device that was returned. A device can for instance become
failed asynchronously.

get_device_medium_pair can also be called after a device has
already been allocated to the queue for a previous request.

We need to manage both cases differently.

When reallocating on the same index, we simply need to remove
the current device from the queue and try to find a new one.

When allocating a request on a queue that was already initialized,
we don't need to do much. We simply recheck the medium's write
compatibility since an admin can for instance lock a medium while
some I/O are scheduled on it.

Since in both cases, there is already a device in the queue, we need
to know whether we are trying to reallocate or not. This is done
by remembering the last index used when calling get_device_medium_pair.

Change-Id: Ia655294db62eddfad285b0a1fec1010919efda5b

Commit:
grouped_write: handle .get_device_medium_pair retries

get_device_medium_pair can be called multiple times with the
same index as long as the upper layer is not satisfied with
the device that was returned. A device can for instance become
failed asynchronously.

get_device_medium_pair can also be called after a device has
already been allocated to the queue for a previous request.

We need to manage both cases differently.

When reallocating on the same index, we simply need to remove
the current device from the queue and try to find a new one.

When allocating a request on a queue that was already initialized,
we don't need to do much. We simply recheck the medium's write
compatibility since an admin can for instance lock a medium while
some I/O are scheduled on it.

Since in both cases, there is already a device in the queue, we need
to know whether we are trying to reallocate or not. This is done
by remembering the last index used when calling get_device_medium_pair.

Change-Id: Ia655294db62eddfad285b0a1fec1010919efda5b

Commit:
grouped_write: make test_lrs_scheduling pass

To make this test pass, a few tests needed to be fixed:
- io_sched_one_request and io_sched_one_request_no_device_available
  where not creating enough devices for the write requests;
- io_sched_4_medium was expecting devices to be returned in order
  but this might not be the case depending on the algorithm.

Change-Id: I022bf2517f1565eca7ef7f8188ea5fc843e1f596

Commit:
grouped_write: make test_lrs_scheduling pass

To make this test pass, a few tests needed to be fixed:
- io_sched_one_request and io_sched_one_request_no_device_available
  where not creating enough devices for the write requests;
- io_sched_4_medium was expecting devices to be returned in order
  but this might not be the case depending on the algorithm.

Change-Id: I022bf2517f1565eca7ef7f8188ea5fc843e1f596

Commit:
grouped_write: document the algoritm's main features

Change-Id: I71ba0d95396237e7fbb963b345c7dca7ed01430a

Commit:
grouped_write: document the algoritm's main features

Change-Id: I71ba0d95396237e7fbb963b345c7dca7ed01430a

Commit:
tests: use grouped_write in acceptance.test

Change-Id: I8cbbc6433996a0ceac8cdf23259821f9a9dc890f

Commit:
tests: use grouped_write in acceptance.test

Change-Id: I8cbbc6433996a0ceac8cdf23259821f9a9dc890f

GauthierEvd commented on src/lrs/io_schedulers/grouped_write.c:
```diff
@@ -0,0 +1,120 @@
+ * vim:expandtab:shiftwidth=4:tabstop=4:
+ */
+/*
+ *  All rights reserved (c) 2014-2022 CEA/DAM.
```
nit: 2025

GauthierEvd commented on src/lrs/io_schedulers/grouped_write.c:
```diff
@@ -32,13 +32,149 @@
+ */
+struct gw_queue {
+  /** Type of layout (e.g. RAID1, RAID4).
+   * For now, this will always be 0 but can be used in the future to
```
question: Why always 0, it will have the same value for RAID1 and RAID4 layout ?

CourrierGui replied:
The idea here is to make it easier in the future if we want to differentiate a RAID4 from a RAID1 repl_count=3 for instance. With the current implementation, they are both the same since n_media = 3. If/when this feature is needed, the only thing that will be required to do is to set the layout_type to the appropriate value.

GauthierEvd commented on src/lrs/io_schedulers/grouped_write.c:
```diff
@@ -133,6 +136,17 @@ struct gw_state {
+
+    grouping = xmalloc(sizeof(*grouping));
+    grouping->name = xstrdup_safe(name);
+    grouping->requests = g_queue_new();
```
question: Can the `g_queue_new` function fail ?

CourrierGui replied:
As mentioned by Sebastien, these functions either don't fail because glib2 asserts that malloc results are not NULL or they fail just in case of memory allocation failures. Since we decided to not care about memory allocation failures in Phobos, I think we can skip the checks here. Worst case senario, phobosd will segfault if any of the glib2's function fail. Which is almost the same behavior as the abort we do in the xmalloc family of functions.

Also keep in mind that by default, malloc's don't fail on Linux. It is when the memory is actually used that the kernel will allocate the physical memory. So what would happen in practice is that malloc would succeed but writing to the allocated memory would produce a page fault which *could* start the OOM killer which *could* kill phobosd.

GauthierEvd commented on src/lrs/io_schedulers/grouped_write.c:
```diff
@@ -198,17 +212,134 @@ static void gw_fini(struct io_scheduler *io_sched)
+        len += strlen(tags[i]) + 1; // + 1 for the ','
+
+    if (n_tags > 0)
+        len--; // remove last +1 for ','
```
question: There is always a coma at end ? It's not `tags1,tags2` ?

SebaGougeaud replied:
I think that's what the lines 249/250 do, removing the last comma.

Maybe you can say below that we allocate one more for the end of string character.

GauthierEvd commented on src/lrs/io_schedulers/grouped_write.c:
```diff
@@ -198,17 +212,134 @@ static void gw_fini(struct io_scheduler *io_sched)
+    queue->n_media = reqc->req->walloc->n_media;
+    queue->devices = xcalloc(queue->n_media, sizeof(*queue->devices));
+    queue->groupings = g_queue_new();
+    queue->grouping_index = g_hash_table_new(g_str_hash, g_str_equal);
```
sugest: check the return value

SebaGougeaud commented on src/lrs/io_schedulers/grouped_write.c:
```diff
@@ -0,0 +1,830 @@
+     */
+    state->allocated = NULL;
+    state->ordered_queues = NULL;
+    state->queues = g_hash_table_new(glib_wqueue_hash, glib_wqueue_equal);
```
nit: IIRC, GLIB initializers do not return NULL pointers, and the documentation do not mention this either

SebaGougeaud commented on src/lrs/io_schedulers/grouped_write.c:
```diff
@@ -0,0 +1,830 @@
+    free(state);
+}
+
+static char *tag2csv(pho_req_write_t *walloc)
```
question: not sure this is relevant, but are tags ordered? are `tag1,tag2`and `tag2,tag1` considered as two different tag lists?

SebaGougeaud replied:
Ok to do this in a specific patch, of course!

SebaGougeaud reviewed (approved):
Seems good to me, great job!

Commit:
grouped_write: sort tags before building the tag list

Tags "foo,bar" and "bar,foo" should be considered the
same when building the tag list in tag2csv. Sort them
before building the list to always be consistent.

Change-Id: I21d1aedca5c996c9778029283d7f13b787965bc3

Commit:
grouped_write: sort tags before building the tag list

Tags "foo,bar" and "bar,foo" should be considered the
same when building the tag list in tag2csv. Sort them
before building the list to always be consistent.

Change-Id: I21d1aedca5c996c9778029283d7f13b787965bc3

Commit:
grouped_write: remove allocation checks for glib2 functions

Change-Id: I04da5a1376bab2f8c115794b75e0e3529551de5c

Commit:
grouped_write: remove allocation checks for glib2 functions

Change-Id: I04da5a1376bab2f8c115794b75e0e3529551de5c

Commit:
tests: add phobosd-client program

This program will allow writing custom scripts to
send specific requests to phobosd. This makes it
possible to write tests that would be otherwise
difficult to do using regular phobos commands.

This program is a simple interpreter that generates
and sends requests to phobosd. For instance:

```
# send a write alloc with 2 media of family tape with tags foo and bar
# The request is sent with the grouping baz
write -t foo,bar -g baz -f tape 2

# wait for a write alloc response for request id 1 (i.e. the write above)
recv 1 write_alloc

# release the media allocated for the first write
release 1

# wait for the write alloc release response
# the release's request ID is 2
recv 2 write_release
```

Request IDs are incremented for each new request. read, write and release
requests will increase the current ID. Therefore, when expecting a
response to a specific request, the ID can be provided as an argument
to the release or recv commands.

Change-Id: Iae10a104b8406a94e87cb4c066ea14e745d47a77

Commit:
tests: add phobosd-client program

This program will allow writing custom scripts to
send specific requests to phobosd. This makes it
possible to write tests that would be otherwise
difficult to do using regular phobos commands.

This program is a simple interpreter that generates
and sends requests to phobosd. For instance:

```
write -t foo,bar -g baz -f tape 2

recv 1 write_alloc

release 1

recv 2 write_release
```

Request IDs are incremented for each new request. read, write and release
requests will increase the current ID. Therefore, when expecting a
response to a specific request, the ID can be provided as an argument
to the release or recv commands.

Change-Id: Iae10a104b8406a94e87cb4c066ea14e745d47a77

SebaGougeaud commented:
I reviewed you last 2 commits, ok for me.

Commit:
grouped_write: keep track of devices currently doing an I/O

Change-Id: I1d5072f3b16af2e3b522e7579d546f8fb2f3004a

Commit:
grouped_write: keep track of devices currently doing an I/O

Change-Id: I1d5072f3b16af2e3b522e7579d546f8fb2f3004a

Commit:
grouped_write: account for ongoing I/O when scheduling new queues

Change-Id: I16118e48258e91f8b1d3aa281a983e0a003d76ac

Commit:
grouped_write: account for ongoing I/O when scheduling new queues

Change-Id: I16118e48258e91f8b1d3aa281a983e0a003d76ac

Commit:
tests: add test_buffered_sync to test_lrs_scheduling

This test makes sure that sync operations are buffered.

Change-Id: If6b1d1c980df7b203db20702b4dfde80e2ccaa57

Commit:
tests: add test_buffered_sync to test_lrs_scheduling

This test makes sure that sync operations are buffered.

Change-Id: If6b1d1c980df7b203db20702b4dfde80e2ccaa57

Commit:
tests: add test_fifo_parallel_write to test_lrs_scheduling

This test makes sure that the FIFO I/O scheduler will schedule
one I/O per device in parallel.

Change-Id: Ibe4a0c01ee65c9c7cf72ddaab887775ce2076167

Commit:
tests: add test_fifo_parallel_write to test_lrs_scheduling

This test makes sure that the FIFO I/O scheduler will schedule
one I/O per device in parallel.

Change-Id: Ibe4a0c01ee65c9c7cf72ddaab887775ce2076167

Commit:
tests: add test_grouped_write_dir to test_lrs_scheduling

This test makes sure that the grouped write I/O scheduler
does not run write I/O in parallel if they share the same
layout, tags and grouping.

Change-Id: I65dd866f0477d02774d38a454b1a6aecac584bd6

Commit:
tests: add test_grouped_write_dir to test_lrs_scheduling

This test makes sure that the grouped write I/O scheduler
does not run write I/O in parallel if they share the same
layout, tags and grouping.

Change-Id: I65dd866f0477d02774d38a454b1a6aecac584bd6

Commit:
phobosd-client: keep initial write request for alloc size

phobosd returns the maximum available size in all the media returned
in the write alloc. We need to set the size initially asked in
the write request when making the corresponding release request.
Add the write allocation requests in a hash table so that we
can retrieve them when making the release request.

Change-Id: Idcec014823e67ea0c758a35e9d1933fb00776301

Commit:
phobosd-client: keep initial write request for alloc size

phobosd returns the maximum available size in all the media returned
in the write alloc. We need to set the size initially asked in
the write request when making the corresponding release request.
Add the write allocation requests in a hash table so that we
can retrieve them when making the release request.

Change-Id: Idcec014823e67ea0c758a35e9d1933fb00776301

GauthierEvd commented on tests/interns/test_lrs_scheduling.c:
comment: It’s not directly related to grouped_write, but the comment saying that the tests expect to receive the media in order makes me think that this test with grouped_read fails if we change the order of the media in the allocation request. It doesn’t match the correct media to the correct device, even though we just mounted them.

CourrierGui replied:
I'm not sure I understand the issue here. The test fixes the order of medium in the allocation request since it is hardcoded here. Also, I modified the test to simply assert the same device is not seen twice. I no longer expect the scheduler to return devices in order and I don't check which medium is mounted where.

GauthierEvd replied:
This is the grouped-read bug where, if the media are already mounted in RAID4 and we do a get, we see that the LRS unmounted two media and swapped them between drives instead of just using them directly.

Basically, when building the request, if you provide the media list as M2, M1, M3, M4 instead of M1, M2, M3, M4, the scheduler returns the pair M2/D1, M1/D2, whereas we should receive M2/D2 and M1/D1 since they are already mounted in a drive.

GauthierEvd commented on tests/externs/cli/test_lrs_scheduling.test:
```diff
@@ -26,6 +26,8 @@ test_dir=$(dirname $(readlink -e $0))
 . $test_dir/utils.sh
 . $test_dir/utils_generation.sh
 
+phobosd_client="$test_dir/utils/.libs/phobosd-client"
```
nitpick: I think it should be in the test_env.sh

GauthierEvd commented on tests/utils/phobosd-client.c:
```diff
@@ -0,0 +1,1153 @@
+
+    for (i = 0; i < n_media; i++) {
+        struct media_info *res;
+        struct dss_filter filter;
```
style: order

GauthierEvd commented on tests/utils/phobosd-client.c:
```diff
@@ -0,0 +1,1153 @@
+}
+
+static void handle_sigterm(int signum)
+{
```
question: Is it normal for this to be empty?

CourrierGui replied:
Unfortunatly, this has been a while so I don't remember why I did this. Assuming I did want to ignore the signal, there is a simpler way with SIG_IGN. My idea at the time was probably to ignore signals since this is a REPL (CTRL-C should not quit the REPL).

GauthierEvd commented on src/lrs/io_schedulers/fifo.c:
```diff
@@ -164,9 +164,31 @@ static int max_write_per_grouping(void)
+}
+
+static bool ongoing_grouping_equal(struct ongoing_grouping *ongoing,
+                             const char *grouping)
```
style: alignment

GauthierEvd commented on src/lrs/io_schedulers/fifo.c:
```diff
@@ -164,9 +164,31 @@ static int max_write_per_grouping(void)
+    if (!grouping)
+        return ongoing->grouping == NULL;
+
+    return !strcmp(ongoing->grouping, grouping);
```
question: You drop the check "ongoing->grouping != NULL", it's because it can never be NULL ?

CourrierGui replied:
That's a good point. It can probably be NULL.

GauthierEvd commented on src/lrs/io_schedulers/grouped_write.c:
```diff
@@ -0,0 +1,888 @@
+    size_t i;
+
+    for (i = 0; i < state->free_devices->len; i++) {
+        struct lrs_dev *device = g_ptr_array_index(state->free_devices, 0);
```
major ? You’re always retrieving index 0, not index i

GauthierEvd commented on tests/utils/requests.c:
```diff
@@ -0,0 +1,182 @@
+    *resps = xmalloc(*n_resps * sizeof(*resps));
+
+    for (i = 0; i < *n_resps; i++)
+        (*resps)[i] = pho_srl_response_unpack(&responses[0].buf);
```
major ? You’re always unpacking the first response `&responses[0]` so the resps array will contain only duplicate copies of the same response.

GauthierEvd commented on src/lrs/io_schedulers/grouped_write.c:
```diff
@@ -0,0 +1,888 @@
+    struct gw_queue *queue;
+
+    /* remove the device from its queue if allocated */
+    queue = g_hash_table_lookup(state->device_to_queue, device);
```
defect: You’re retrieving queue from this hashtable, but I don’t see anywhere that queue is being added to it. I think you need to do that in gw_get_device_medium_pair.

GauthierEvd commented on tests/utils/requests.c:
```diff
@@ -0,0 +1,182 @@
+                              struct string_array *tags,
+                              const char *library)
+{
+    size_t *n_tags = xmalloc(tags->count * sizeof(*n_tags));
```
defect: it should be `xmalloc(n_media * sizeof(*n_tags))`

GauthierEvd commented on tests/utils/phobosd-client.c:
```diff
@@ -0,0 +1,1173 @@
+
+    write = make_write_request(ctxt->last_id++, n_media, size,
+                               family, grouping, &tags, library);
+    g_hash_table_insert(ctxt->write_reqs_by_id, &write->id, write);
```
minor: The requests are never removed from the hash table, but they are freed when a release is received, so the requests stored in the hash table are no longer valid.

GauthierEvd commented on tests/utils/phobosd-client.c:
```diff
@@ -0,0 +1,1173 @@
+    char c;
+    int rc;
+
+    while ((c = getopt_long(argc, argv, "t:g:", options, NULL)) != -1) {
```
major: You forgot to add the library and family options

GauthierEvd commented on tests/utils/phobosd-client.c:
```diff
@@ -0,0 +1,1173 @@
+                     int argc, char **argv)
+{
+    struct option options[] = {
+        {"grouping",    no_argument,       0,  'g'},
```
major: it should be `required_argument ` here

GauthierEvd commented on src/lrs/io_schedulers/grouped_write.c:
```diff
@@ -0,0 +1,888 @@
+
+new_grouping:
+    grouping = gw_grouping_new(name);
+    g_queue_push_head(queue->groupings, grouping);
```
major: Here, you insert the new grouping at the beginning of the list, and then when handling peek request, you also take the first grouping. As a result, we always process the most recently received grouping first, which means we lose the arrival order of the groupings. That contradicts the comment on line 91.

GauthierEvd commented on tests/utils/phobosd-client.c:
```diff
@@ -0,0 +1,1173 @@
+        return rc;
+
+    read = make_read_request(ctxt->last_id++, n_media, n_required,
+                             PHO_RSC_DIR, (const char **)media,
```
minor: It should be the family retrieved just before; otherwise, we might run into an issue if the media is a tape.

Commit:
grouped_write: various bug fixes

grouped_write:
- check whether ld_ongoing_grouping is NULL before strcmp
- push new grouping at the end of the queue's grouping list
  to preserve the order of arrival
- insert queue in device_to_queue hashtable when a queue is
  first allocated

phobosd-client:
- fix write command argument parsing
- use the right family in read allocations
- remove write requests from write_reqs_by_id hashtable after the
  release is received

Change-Id: I6feafd82bc95872a7a5b6441f328020be9e9119d
